runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/FlinkStateInternalsTest.java

```diff
@@ -190,29 +206,224 @@ public void testMergeBagIntoNewNamespace() throws Exception {
     StateMerging.mergeBags(Arrays.asList(bag1, bag2, bag3), bag3);
 
     // Reading the merged bag gets both the contents
-    assertThat(bag3.read(), Matchers.containsInAnyOrder("Hello", "World", "!"));
+    assertThat(bag3.read(), containsInAnyOrder("Hello", "World", "!"));
     assertThat(bag1.read(), Matchers.emptyIterable());
     assertThat(bag2.read(), Matchers.emptyIterable());
   }
 
   @Test
+  public void testSet() throws Exception {
+    SetState<String> value = underTest.state(NAMESPACE_1, STRING_SET_ADDR);
+
+    // State instances are cached, but depend on the namespace.
+    assertThat(value, equalTo(underTest.state(NAMESPACE_1, STRING_SET_ADDR)));
+    assertThat(value, not(equalTo(underTest.state(NAMESPACE_2, STRING_SET_ADDR))));
+
+    // empty
+    assertThat(value.read(), Matchers.emptyIterable());
+    assertFalse(value.contains("A").read());
+
+    // add
+    value.add("A");
+    value.add("B");
+    value.add("A");
+    assertFalse(value.addIfAbsent("B").read());
```

Comment:
Yes, good point, another similar situation is: 
```
//At first, there is no "B"
ReadableState<Boolean> readable = value.contains("B");
value.add("B");
Boolean result = readable.read();
```
The in-memory state will return false while the flink state will return true.
@kennknowles what do you think about these two cases?